File: server/java/render/src/main/java/com/metaobjects/render/recover/Strip.java
package com.metaobjects.render.recover;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Stage 1: remove markdown code-fence markers. Prose around the payload is left for Locate. */
public final class Strip {
    private Strip() {}

    // Captures the body inside a fenced block; optional language tag (json/xml/etc) is dropped.
    private static final Pattern FENCE = Pattern.compile(
            "```[a-zA-Z0-9_-]*\\s*\\r?\\n(.*?)\\r?\\n?```",
            Pattern.DOTALL);

    public static String strip(String raw) {
        if (raw == null) return "";
        Matcher m = FENCE.matcher(raw);
        if (m.find()) {
            StringBuilder sb = new StringBuilder();
            sb.append(raw, 0, m.start());
            sb.append(m.group(1));
            sb.append(raw.substring(m.end()));
            return sb.toString().trim();
        }
        return raw.trim();
    }
}

File: server/python/src/metaobjects/render/recover/strip.py
"""Stage 1: remove markdown code-fence markers.

Prose around the payload is left for ``locate`` to handle.
"""
from __future__ import annotations

import re

# Captures the body inside a fenced block; optional language tag (json/xml/etc) is dropped.
_FENCE = re.compile(r"```[a-zA-Z0-9_-]*[ \t]*\r?\n(.*?)\r?\n?```", re.DOTALL)


def strip(raw: str | None) -> str:
    """Strip the first markdown code fence, splicing its body back in place; else trim."""
    if raw is None:
        return ""
    m = _FENCE.search(raw)
    if m:
        return (raw[: m.start()] + m.group(1) + raw[m.end():]).strip()
    return raw.strip()
